File: in_Java/0103 Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        if(root == null) return res;
        q.add(root);
        TreeNode cur;
        int len;
        int id = 0;
        List<Integer> temp;
        while(q.size() != 0){
            len = q.size();
            temp = new ArrayList<>();
            while(len != 0){
                cur = q.poll();
                if(id % 2 == 0){
                    temp.add(cur.val);
                }else{
                    temp.add(0,cur.val);
                }
                if(cur.left != null) q.add(cur.left);
                if(cur.right != null) q.add(cur.right);
                len--;
            }
            id++;
            res.add(temp);
        }
        return res;
    }
}

File: in_Python/0103 Binary Tree Zigzag Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        q = []
        res = []
        idx = 0
        if root == None:
            return res
        q.append(root)
        while len(q) != 0:
            n = len(q)
            cur = []
            for i in range(n):
                head = q.pop(0)
                if head.left != None:
                    q.append(head.left)
                if head.right != None:
                    q.append(head.right)
                if idx == 0:
                    cur.append(head.val)
                else:
                    cur.insert(0, head.val)
            if idx == 0:
                idx = 1
            else:
                idx = 0
            res.append(cur)
        return res